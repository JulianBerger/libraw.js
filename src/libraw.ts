const binary = require('node-pre-gyp');
const path = require('path');
const bindingPath = binary.find(path.resolve(path.join(__dirname, '../package.json')));
const librawAddon = require(bindingPath);

interface LibRawWrapper {
  getMetadata: () => { [key: string]: unknown };
  getThumbnail: () => Buffer;
  getXmp: () => Buffer;
}

/**
 * Wraps LibRaw's functionality.
 */
export class LibRaw {
  private libraw: LibRawWrapper;

  constructor() {
    this.libraw = null;
  }

  /**
   * Initializes a LibRaw instance from a file in
   * the specified path.
   * @param filePath path to the RAW file to process
   */
  readFile(filePath: string): Promise<void> {
    return this.accessLibRaw<void>(() => {
      this.libraw = new librawAddon.LibRawWrapper(filePath);
    });
  }

  /**
   * Initializes a LibRaw instance from a file in memory.
   * @param buffer the RAW file data
   */
  readBuffer(buffer: Buffer) {
    return this.accessLibRaw<void>(() => {
      this.libraw = new librawAddon.LibRawWrapper(buffer);
    });
  }

  /**
   * Returns an object containing the RAW metadata.
   */
  getMetadata(): Promise<{ [key: string]: unknown }> {
    return this.accessLibRaw(() => this.libraw.getMetadata());
  }

  /**
   * Returns the XMP data of the RAW file.
   */
  getXmp(): Promise<Buffer> {
    return this.accessLibRaw(() => this.libraw.getXmp());
  }

  /**
   * Unpacks and returns the bytes for the image's thumbnail.
   */
  getThumbnail(): Promise<Buffer> {
    return this.accessLibRaw(() => this.libraw.getThumbnail());
  }

  /**
   * Abstracts interactions with LibRaw to avoid boilerplate
   * async code repetitions in public methods.
   * @param executor the interaction with LibRaw
   */
  private accessLibRaw<T>(executor): Promise<T> {
    return new Promise((resolve, reject) => {
      try {
        resolve(executor());
      } catch (e) {
        reject(e);
      }
    });
  }
}
