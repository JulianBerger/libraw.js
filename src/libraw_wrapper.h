#include <napi.h>
#include "libraw/libraw.h"

class LibRawWrapper: public Napi::ObjectWrap<LibRawWrapper> {
  public:
    static Napi::Object Init(Napi::Env& env, Napi::Object& exports);
    LibRawWrapper(const Napi::CallbackInfo& info);
    ~LibRawWrapper();
    Napi::Value CameraCount(const Napi::CallbackInfo& info);
    Napi::Value CameraList(const Napi::CallbackInfo& info);
    Napi::Value GetMetadata(const Napi::CallbackInfo& info);
    Napi::Value GetThumbnail(const Napi::CallbackInfo& info);
    Napi::Value GetXmpData(const Napi::CallbackInfo& info);
    Napi::Value OpenFile(const Napi::CallbackInfo& info);
    Napi::Value OpenBuffer(const Napi::CallbackInfo& info);
    Napi::Value Unpack(const Napi::CallbackInfo& info);
    Napi::Value UnpackThumb(const Napi::CallbackInfo& info);
    Napi::Value ErrorCount(const Napi::CallbackInfo& info);
    Napi::Value Version(const Napi::CallbackInfo& info);
    Napi::Value VersionNumber(const Napi::CallbackInfo& info);
    Napi::Value StrError(const Napi::CallbackInfo& info);
    void RecycleDatastream(const Napi::CallbackInfo& info);
    void Recycle(const Napi::CallbackInfo& info);
  private:
    LibRaw* processor_;
};