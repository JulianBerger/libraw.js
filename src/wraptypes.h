#include "libraw/libraw.h"

Napi::Value WrapLibRawData(Napi::Env* env, libraw_data_t* data);
