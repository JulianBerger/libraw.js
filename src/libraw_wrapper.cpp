#include <napi.h>
#include "libraw_wrapper.h"
#include "wraptypes.h"
#include <fstream>

Napi::Object LibRawWrapper::Init(Napi::Env& env, Napi::Object& exports) {
  Napi::HandleScope scope(env);

  Napi::Function func =
    DefineClass(
      env,
      "LibRawWrapper",
      {
        InstanceMethod("getMetadata", &LibRawWrapper::GetMetadata),
        InstanceMethod("getThumbnail", &LibRawWrapper::GetThumbnail),
        InstanceMethod("getXmp", &LibRawWrapper::GetXmpData)
      }
    );

  exports.Set("LibRawWrapper", func);
  return exports;
}

LibRawWrapper::LibRawWrapper(const Napi::CallbackInfo& info): Napi::ObjectWrap<LibRawWrapper>(info) {
  Napi::Env env = info.Env();
  Napi::HandleScope scope(env);
  if (info.Length() < 1)
  {
    Napi::TypeError::New(env, "Expected path to image").ThrowAsJavaScriptException();
    return;
  }
  if (!info[0].IsString() && (!info[0].IsBuffer() && info.Length() > 1))
  {
    Napi::TypeError::New(env, "Expected path to image or buffer with size").ThrowAsJavaScriptException();
    return;
  }

  if (info[0].IsString())
  {
    Napi::String path = info[0].As<Napi::String>();
    std::fstream f(path.Utf8Value().c_str());
    if (!f.good())
    {
      Napi::TypeError::New(env, "File does not exist").ThrowAsJavaScriptException();
      return;
    }
    this->processor_ = new LibRaw();
    this->processor_->open_file(path.Utf8Value().c_str());
  }
  else if (info[0].IsBuffer() && info[1].IsNumber())
  {
    this->processor_ = new LibRaw();
    Napi::Buffer<char> buff = info[0].As<Napi::Buffer<char>>();
    this->processor_->open_buffer(buff.Data(), buff.Length());
  }
  else
  {
    Napi::TypeError::New(env, "LibRaw constructor did not receive required parameters.").ThrowAsJavaScriptException();
  }
}

Napi::Value LibRawWrapper::GetThumbnail(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  Napi::Object o = Napi::Object::New(env);
  
  this->processor_->unpack_thumb();
  if (this->processor_->imgdata.thumbnail.thumb)
  {
    Napi::Buffer<char> b = Napi::Buffer<char>::New(
      env,
      this->processor_->imgdata.thumbnail.thumb,
      this->processor_->imgdata.thumbnail.tlength
    );
    return b;
  }

  return o;
}

Napi::Value LibRawWrapper::GetXmpData(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  Napi::Object o = Napi::Object::New(env);

  char *xmp = this->processor_->imgdata.idata.xmpdata;
  if (xmp)
  {
    Napi::Buffer<char> b = Napi::Buffer<char>::New(
      env,
      xmp,
      this->processor_->imgdata.idata.xmplen
    );
    return b;
  }
  return o;
}

Napi::Value LibRawWrapper::GetMetadata(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  libraw_data_t data = this->processor_->imgdata;
  return WrapLibRawData(&env, &data);
}

LibRawWrapper::~LibRawWrapper()
{
  delete this->processor_;
}
