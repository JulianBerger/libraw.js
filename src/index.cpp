#include "libraw_wrapper.h"

Napi::Object Init(Napi::Env env, Napi::Object exports)
{
  return LibRawWrapper::Init(env, exports);
}

NODE_API_MODULE(NODE_GYP_MODULE_NAME, Init)
