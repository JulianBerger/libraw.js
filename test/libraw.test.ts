import { LibRaw } from '../src/libraw';
import path from 'path';
import fs from 'fs';

describe('LibRaw', () => {
  test('throws exception for non-existent file', () => {
    const lr = new LibRaw();
    expect(lr.readFile('./some/non-existent-path')).rejects.toEqual(
      new Error('File does not exist')
    );
  });

  test('basic metadata fields supported', async () => {
    expect.assertions(3);
    const lr = new LibRaw();
    await lr.readFile(
      path.join(__dirname, 'test_images/RAW_SONY_ILCA-77M2.ARW')
    );
    const metadata = await lr.getMetadata();
    expect(metadata.idata).toMatchSnapshot();
    expect(metadata.lens).toMatchSnapshot();
    expect(metadata.other).toMatchSnapshot();
  });

  test('reads img data from buffer', async () => {
    expect.assertions(3);
    const lr = new LibRaw();
    const buffer = fs.readFileSync(
      path.join(__dirname, 'test_images/RAW_SONY_ILCA-77M2.ARW')
    );
    await lr.readBuffer(buffer);
    const metadata = await lr.getMetadata();
    expect(metadata.idata).toMatchSnapshot();
    expect(metadata.lens).toMatchSnapshot();
    expect(metadata.other).toMatchSnapshot();
  });

  test('parses xmp', async () => {
    const lr = new LibRaw();
    await lr.readFile(path.join(__dirname, 'test_images/RAW_NIKON_Z6.NEF'));
    const xmp = (await lr.getXmp()).toString('utf8');
    expect(xmp).toMatchSnapshot();
  });
});
