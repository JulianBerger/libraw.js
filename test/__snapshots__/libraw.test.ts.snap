// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LibRaw basic metadata fields supported 1`] = `
Object {
  "cdesc": "RGBG",
  "colors": 3,
  "dng_version": 0,
  "filters": 3031741620,
  "guard": "",
  "is_foveon": 0,
  "make": "Sony",
  "model": "ILCA-77M2",
  "raw_count": 1,
  "software": "ILCA-77M2 v1.01",
  "xmplen": 0,
  "xtrans": Array [
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
    ],
  ],
  "xtrans_abs": Array [
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
    ],
  ],
}
`;

exports[`LibRaw basic metadata fields supported 2`] = `
Object {
  "EXIF_MaxAp": 1.799575,
  "FocalLengthIn35mmFormat": 52,
  "InternalLensSerial": "",
  "Lens": "DT 35mm F1.8 SAM",
  "LensMake": "",
  "LensSerial": "",
  "MaxAp4MaxFocal": 1.8,
  "MaxAp4MinFocal": 1.8,
  "MaxFocal": 35,
  "MinFocal": 35,
  "dng": Object {
    "MaxAp4MaxFocal": 0,
    "MaxAp4MinFocal": 0,
    "MaxFocal": 0,
    "MinFocal": 0,
  },
  "makernotes": Object {
    "Adapter": "",
    "AdapterID": 0,
    "Attachment": "",
    "AttachmentID": 0,
    "CamID": 319,
    "CameraFormat": 1,
    "CameraMount": 1,
    "CurAp": 5.656854,
    "CurFocal": 0,
    "FocalLengthIn35mmFormat": 0,
    "FocalType": 0,
    "FocusRangeIndex": 0,
    "Lens": "",
    "LensFStops": 0,
    "LensFeatures_pre": "",
    "LensFeatures_suf": "",
    "LensFormat": 1,
    "LensID": "",
    "LensMount": 1,
    "MaxAp": 0,
    "MaxAp4CurFocal": 1.8,
    "MaxAp4MaxFocal": 0,
    "MaxAp4MinFocal": 1.8,
    "MaxFocal": 0,
    "MinAp": 0,
    "MinAp4CurFocal": 22.200001,
    "MinAp4MaxFocal": 0,
    "MinAp4MinFocal": 0,
    "MinFocal": 35,
    "MinFocusDistance": 0,
    "Teleconverter": "",
    "TeleconverterID": 0,
    "body": "",
  },
  "nikon": Object {
    "NikonEffectiveMaxAp": 0,
    "NikonLensFStops": 0,
    "NikonLensIDNumber": 0,
    "NikonLensType": 0,
    "NikonMCUVersion": 0,
  },
}
`;

exports[`LibRaw basic metadata fields supported 3`] = `
Object {
  "AmbientTemperature": -1000,
  "BatteryTemperature": 29.444445,
  "CameraTemperature": -1000,
  "FlashEC": 0,
  "FlashGN": 0,
  "LensTemperature": -1000,
  "SensorTemperature": -1000,
  "SensorTemperature2": -1000,
  "aperture": 5.6,
  "artist": "",
  "desc": "                               ",
  "exifAcceleration": 0,
  "exifAmbientTemperature": -1000,
  "exifCameraElevationAngle": 0,
  "exifHumidity": 0,
  "exifPressure": 0,
  "focal_len": 35,
  "gpsdata": Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  "iso_speed": 100,
  "parsed_gps": Object {
    "altitude": 0,
    "altref": 0,
    "gpsparsed": 0,
    "gpsstatus": 0,
    "gpstimestamp": Array [
      0,
      0,
      0,
    ],
    "latitude": Array [
      0,
      0,
      0,
    ],
    "latref": 0,
    "longref": 0,
    "longtitude": Array [
      0,
      0,
      0,
    ],
  },
  "real_ISO": 100,
  "shot_order": 0,
  "shutter": 0.0125,
  "timestamp": 1414528361,
}
`;

exports[`LibRaw parses xmp 1`] = `"<?xpacket begin=\\"ï»¿\\" id=\\"W5M0MpCehiHzreSzNTczkc9d\\"?><x:xmpmeta xmlns:x=\\"adobe:ns:meta/\\"><rdf:RDF xmlns:rdf=\\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\\"><rdf:Description rdf:about=\\"\\" xmlns:xmp=\\"http://ns.adobe.com/xap/1.0/\\"><xmp:CreatorTool>NIKON Z 6 Ver.02.00     </xmp:CreatorTool><xmp:CreateDate>2019-07-26T20:36:57.90</xmp:CreateDate><xmp:Rating>0</xmp:Rating> </rdf:Description><rdf:Description rdf:about=\\"\\" xmlns:crd=\\"http://ns.adobe.com/camera-raw-defaults/1.0/\\"><crd:Exposure2012>0.00</crd:Exposure2012><crd:Highlights2012>0</crd:Highlights2012><crd:Shadows2012>0</crd:Shadows2012><crd:LuminanceSmoothing>39</crd:LuminanceSmoothing><crd:LuminanceNoiseReductionDetail>75</crd:LuminanceNoiseReductionDetail><crd:LuminanceNoiseReductionContrast>0</crd:LuminanceNoiseReductionContrast><crd:ColorNoiseReduction>10</crd:ColorNoiseReduction><crd:ColorNoiseReductionDetail>50</crd:ColorNoiseReductionDetail><crd:ColorNoiseReductionSmoothness>50</crd:ColorNoiseReductionSmoothness><crd:Sharpness>40</crd:Sharpness><crd:SharpenRadius>2.00</crd:SharpenRadius><crd:SharpenDetail>25</crd:SharpenDetail><crd:SharpenEdgeMasking>0</crd:SharpenEdgeMasking><crd:Contrast2012>0</crd:Contrast2012><crd:Saturation>0</crd:Saturation><crd:CameraProfile>Camera Standard</crd:CameraProfile></rdf:Description></rdf:RDF></x:xmpmeta>                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           <?xpacket end=\\"w\\"?> "`;

exports[`LibRaw reads img data from buffer 1`] = `
Object {
  "cdesc": "RGBG",
  "colors": 3,
  "dng_version": 0,
  "filters": 3031741620,
  "guard": "",
  "is_foveon": 0,
  "make": "Sony",
  "model": "ILCA-77M2",
  "raw_count": 1,
  "software": "ILCA-77M2 v1.01",
  "xmplen": 0,
  "xtrans": Array [
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
    ],
  ],
  "xtrans_abs": Array [
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
    ],
  ],
}
`;

exports[`LibRaw reads img data from buffer 2`] = `
Object {
  "EXIF_MaxAp": 1.799575,
  "FocalLengthIn35mmFormat": 52,
  "InternalLensSerial": "",
  "Lens": "DT 35mm F1.8 SAM",
  "LensMake": "",
  "LensSerial": "",
  "MaxAp4MaxFocal": 1.8,
  "MaxAp4MinFocal": 1.8,
  "MaxFocal": 35,
  "MinFocal": 35,
  "dng": Object {
    "MaxAp4MaxFocal": 0,
    "MaxAp4MinFocal": 0,
    "MaxFocal": 0,
    "MinFocal": 0,
  },
  "makernotes": Object {
    "Adapter": "",
    "AdapterID": 0,
    "Attachment": "",
    "AttachmentID": 0,
    "CamID": 319,
    "CameraFormat": 1,
    "CameraMount": 1,
    "CurAp": 5.656854,
    "CurFocal": 0,
    "FocalLengthIn35mmFormat": 0,
    "FocalType": 0,
    "FocusRangeIndex": 0,
    "Lens": "",
    "LensFStops": 0,
    "LensFeatures_pre": "",
    "LensFeatures_suf": "",
    "LensFormat": 1,
    "LensID": "",
    "LensMount": 1,
    "MaxAp": 0,
    "MaxAp4CurFocal": 1.8,
    "MaxAp4MaxFocal": 0,
    "MaxAp4MinFocal": 1.8,
    "MaxFocal": 0,
    "MinAp": 0,
    "MinAp4CurFocal": 22.200001,
    "MinAp4MaxFocal": 0,
    "MinAp4MinFocal": 0,
    "MinFocal": 35,
    "MinFocusDistance": 0,
    "Teleconverter": "",
    "TeleconverterID": 0,
    "body": "",
  },
  "nikon": Object {
    "NikonEffectiveMaxAp": 0,
    "NikonLensFStops": 0,
    "NikonLensIDNumber": 0,
    "NikonLensType": 0,
    "NikonMCUVersion": 0,
  },
}
`;

exports[`LibRaw reads img data from buffer 3`] = `
Object {
  "AmbientTemperature": -1000,
  "BatteryTemperature": 29.444445,
  "CameraTemperature": -1000,
  "FlashEC": 0,
  "FlashGN": 0,
  "LensTemperature": -1000,
  "SensorTemperature": -1000,
  "SensorTemperature2": -1000,
  "aperture": 5.6,
  "artist": "",
  "desc": "                               ",
  "exifAcceleration": 0,
  "exifAmbientTemperature": -1000,
  "exifCameraElevationAngle": 0,
  "exifHumidity": 0,
  "exifPressure": 0,
  "focal_len": 35,
  "gpsdata": Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  "iso_speed": 100,
  "parsed_gps": Object {
    "altitude": 0,
    "altref": 0,
    "gpsparsed": 0,
    "gpsstatus": 0,
    "gpstimestamp": Array [
      0,
      0,
      0,
    ],
    "latitude": Array [
      0,
      0,
      0,
    ],
    "latref": 0,
    "longref": 0,
    "longtitude": Array [
      0,
      0,
      0,
    ],
  },
  "real_ISO": 100,
  "shot_order": 0,
  "shutter": 0.0125,
  "timestamp": 1414528361,
}
`;
